data/shared/citizen/scripting/v8/main.js

```diff
@@ -1,156 +1,172 @@
+	 * @param {int} ref
+	 */
+	Citizen.setDeleteRefFunction(function(ref) {
+		refFunctionsMap.delete(ref);
+	});
+
+	/**
+	 * Invokes ref function
+	 * 
+	 * @param {int} ref 
+	 * @param {UInt8Array} args 
+	 */
+	Citizen.setCallRefFunction(function(ref, argsSerialized) {
+		if (!refFunctionsMap.has(ref)) {
+			console.error('Invalid ref call attempt:', ref);
+
+			return pack({});
 		}
 
-		requestAnimationFrame(cb)
-		{
-			this.animationFrames.push(cb);
+		return pack(refFunctionsMap.get(ref)(unpack(argsSerialized)));
+	});
+
+	/**
+	 * Duplicates ref function
+	 * 
+	 * @param {int} ref
+	 */
+	Citizen.setDuplicateRefFunction(function(ref) {
+		if (refFunctionsMap.has(ref)) {
+			const refFunction = refFunctionsMap.get(ref);
+			const newRef = nextRefIdx();
+
+			refFunctionsMap.set(newRef, refFunction);
+
+			return newRef;
 		}
 
-		tick()
-		{
-			// process timers
-			let currentTime = Citizen.getTickCount();
-
-			for (let timerID in this.timers)
-			{
-				let timer = this.timers[timerID];
-
-				if (currentTime >= (timer.lastRun + timer.interval))
-				{
-					try
-					{
-						timer.callback();
-					}
-					catch (e)
-					{
-						// TODO: replace with console.log once implemented
-						print(e);
-						print(e.stack);
-					}
-
-					timer.lastRun = currentTime;
-				}
-			}
+		return -1;
+	});
 
-			// process animation frames
-			for (let animationFrame of this.animationFrames)
-			{
-				try
-				{
-					animationFrame();
-				}
-				catch (e)
-				{
-					// TODO: replace with console.log once implemented
-					print(e);
-					print(e.stack);
-				}
-			}
+	// Events
+	const emitter = new EventEmitter2();
+	const rawEmitter = new EventEmitter2();
+	const netSafeEventsNames = new Set(['playerDropped', 'playerConnecting']);
 
-			this.animationFrames = [];
+	// Raw events
+	global.addRawEventListener = rawEmitter.on.bind(rawEmitter);
+
+	// Client events
+	global.addEventListener = (name, callback, netSafe = false) => {
+		if (netSafe) {
+			netSafeEventsNames.add(name);
 		}
+
+		emitter.on(name, callback);
 	};
+	global.on = global.addEventListener;
 
-	// instantiate the timer manager
-	let timerManager = new TimerManager();
+	// Net events
+	global.addNetEventListener = (name, callback) => global.addEventListener(name, callback, true);
+	global.onNet = global.addNetEventListener;
 
-	// set global functions
-	window.setTimeout = timerManager.setTimeout.bind(timerManager);
-	window.clearTimeout = timerManager.clearTimeout.bind(timerManager);
+	global.removeEventListener = emitter.off.bind(emitter);
 
-	window.setInterval = timerManager.setInterval.bind(timerManager);
-	window.clearInterval = timerManager.clearInterval.bind(timerManager);
+	global.emit = (name, ...args) => {
+		const dataSerialized = pack(data);
 
-	window.requestAnimationFrame = timerManager.requestAnimationFrame.bind(timerManager);
+		//                    TRIGGER_EVENT_INTERNAL
+		Citizen.invokeNative("0x91310870", dataSerialized, dataSerialized.length);
```

Comment:
If a natives_* is guaranteed to be loaded, should this not invoke the native wrapper?